### PR TITLE
Resolve References using cache for Product resource.

### PR DESCRIPTION
**Pull request by praveenkumarct**
- Create a service which accepts client and map(cache) and products to transform and map.
- update utils to perform reference resolution using cache.
- Unit and Integration tests.
- Proper Javadocs.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/products/service/ProductReferenceTransformService.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/ProductReferenceTransformService.java:18]:
```suggestion
   * <p>If the reference ids are already cached, key values are pulled from the cache, otherwise it
```

File: src/main/java/com/commercetools/sync/products/service/ProductReferenceTransformService.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/ProductReferenceTransformService.java:22]:
```suggestion
   * <p>Then maps the Product to ProductDraft by performing reference resolution considering idToKey
```

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:52]: We don't have a logger in the java-sync 🤔

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:75]: Looks like this error handling part is more related to the project-sync code, not related to java-sync . In this case, we might throw the error from the service as we could not resolve this right? The library user should decide what to do.

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:1]: What do you think about moving the services to the "com.commercetools.sync.services.transform" package?

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:60]: What about providing a default map backed by caffeeine as a reference? So the library users would use it without investing in their best practices?

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:480]: can you put withTotal=false and a limit to the query, otherwise it's limited to 20 custom objects, (we figured that with benchmark test in the previous release if you remember)

File: src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java:234]: That would conflict with Andi's work, please consider not merging those changes into master until the whole development is done.

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java:43]: Is this comment required?

**Review by ahmetoz — CHANGES_REQUESTED** [review]

File: src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java:234]: Sure, I will wait for Andi to merge his changes into master and then i will fix the conflicts in my branch.

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/db6bca046d69046ac3d92b2b3bac89e65ac5f1f4

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/c5e5c9b9c226f383f5d8ba70db7357c7bfdd0d8d

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:1]: Since the class name already contains "Transform" and also we don't have any other service classes as of now. I think it is good to have in com.commercetools.sync.service package itself. Let me know if you think still it is better to have transform package!

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:60]: Do you mean to provide the default map when the map is null ?  Or just update the javadocs about the default cache Caffeine for library users to implement it with example?

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/c15173a3fe3479816c7abb1d34ecc09a498b8f0e

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/c15173a3fe3479816c7abb1d34ecc09a498b8f0e

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:54]: I'm asking giving a reference class for to use by default on our snippets, also document for custom use cases, and invalidation scenarios for our clients (in the doc snippets)

For instance we would have a class called `DefaultTransformServiceCache` (or you name it) which is documented as an example in the docs and used by integration tests.

Class is backed by caffeine cache with LRU cache eviction strategy, which is taking the size and documents the best practices for the caching needs. It could also have additional utility methods like invalidateAll method, or others.

**Review by ahmetoz — CHANGES_REQUESTED** [review]

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:54]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/8a458e56c6e4a63b346644e26f27b6cf44057a5d
I have added Default cache using Caffeine and this will be documented as part of documentation task.

**Comment by praveenkumarct** [comment]
@ahmetoz 
When I used a Default cache(Caffeine), observed that it doesn't allow null values into the map.
Because of this, Integration tests failed for the case(where ProductType key value is null). To handle this case, I have introduced a new static variable (BLANK_KEY_VALUE = "KEY_VALUE_EITHER_NULL_OR_BLANK") in BaseTransformServiceImpl.
Let me know your view on this? Do you have any other clean way of handling this scenario?

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:54]: First it's not something related to product package, maybe this should be part of the utils, also instead of explaining it, I pushed a commit to another branch, which I meant in my comment: https://github.com/commercetools/commercetools-sync-java/commit/c258f2cbe171378379d6fbc0ef44659a84f64d54

Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:53]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/840a2dea62f16b192af58dbc402883551f0a39f5 merged into this branch using cherry-pick

File: src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java:235]: you might base to `release_5` branch, that Andi created already, so we could review this again, as there would be lots of conflict changes

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java:31]:
```suggestion
  public static final String KEY_IS_NOT_SET_PLACE_HOLDER = "KEY_IS_NOT_SET";
```

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java:84]: Here we need to filter out the keys not exist but have a placeholder, KEY_IS_NOT_SET_PLACE_HOLDER. 
Otherwise, when the key is added to the project it would not be synced and the exceptions will not be resolved.

Could you check for other code if `getNonCachedReferenceIds` is overridden in subclasses? Also please try to check document reference for custom objects, as it's not stored as a simple key but the combination of container and key (("id", "co-container|co-key"))

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:414]: Looks like, it's fetching the keys that is already in the cache: 

```suggestion
      return fetchCustomObjectKeys(nonCachedCustomObjectIds);
```

File: src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java:468]: I checked if this would be needed a placeholder here, but in commercetools, both container and key are required fields, looks like we are safe in this case, so looks like no additional check is needed.

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (ahmetoz) [inline src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java:94]: Here as you were doing before, you could keep the placeholder as a key if it could be fetched from CTP, then during reference resolution, it could use the same error handling, which is also failing fast to error callback before executing actions.

**Review by ahmetoz — COMMENTED** [review]
As requested, I added my comments considering not set keys on the target project.

File: src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/products/utils/ProductReferenceResolutionUtils.java:236]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/4890668f7c7cf19bfab7e7ed41860fdbc8a67b3f

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/4c3892041cb3afa8f9a6167c31647b69aebf75cf

File: src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java
Comment (praveenkumarct) [reply src/main/java/com/commercetools/sync/services/impl/BaseTransformServiceImpl.java:78]: https://github.com/commercetools/commercetools-sync-java/pull/697/commits/ffe6f19db27aa206410807c5e56943283370b0c3
I am checking for custom Objects.

**Comment by praveenkumarct** [comment]
> it will be always not blank, so this check might be unnecessary.

https://github.com/commercetools/commercetools-sync-java/pull/697/commits/ffe6f19db27aa206410807c5e56943283370b0c3

**Review by ahmetoz — DISMISSED** [review]

**Pull request merged**
